### PR TITLE
Heavily Optimized `std.mem.eql` with SIMD

### PR DESCRIPTION
**Pull request by Rexicon226**
The current stdlib implementation is not as good as it can be for `u8`, which is the most commonly used one by far. @kprotty and I have taken the task of vectorizing and accelerating this function by orders of magnitude. The idea here is to split this api into 2 different functions. The normal `eql` function, which works on anything, and the much more optimized `eqlBytes`, which specifically optimizes bytes. `eql` calls `eqlBytes` if `T` is `u8`.

For any other integers, we still use a much faster xor accumulator design, see benchmarks below.

## Benchmarks

The benchmarks are run in as good of a script as I was able to make, including using cpu counters, pinning the process to 1 core, and clearing caches between different function runs. Warmups happen for each function, as well as at least 1000 runs.

The CPU the benchmarks were run on:
```
AMD Ryzen 9 6900HS Creator Edition (16) @ 3.293GHz
```

Important to note that the cpu does have `constant_tsc` enabled.

There are 3 benchmarks for the 3 categories, `start`, `middle`, `same`.

`start` is when the difference is at the first byte. This is meant to show the start-up cost, and how fast it exists.
`middle` shows a sort of "middle" case. Is it able to find it faster in smaller inputs? Does it have an early exit?
`same` is the worst case scenario. Here the entire input must be checked, and is actually the most common input. People will usually use equality to see it's true rather than false.

Run with: `zig build-exe benchmark.zig -lc -OReleaseFast`
<details>
  <summary>benchmarks</summary>

plot

</details>

The `start` on both versions is measured under 100 cycles, which I consider to be starting to get close to the error margin. At this level, there isn't any point to comparing them, and both running this fast will have virtually no performance difference for the user.

Another benchmark I tried was running the `perf_test.zig` inside of `lib/std/zig`. I felt that the parser and tokenizer would be good spot with many `mem.eql` usages.

| stdlib | xor eql |
|--------|--------|
| `parsing speed: 117.96MiB/s` | `parsing speed: 124.25MiB/s` | 

that is a solid 6% speed increase on average. Note that I up the iteration count from 100 to 10000, to get the most precise readings.

I also added `sched_setaffinity` , as it seems the `getaffinity` was present, but there was no `setaffinity` and I needed it for the benchmarking.

### CODE REVIEW

File: lib/std/mem.zig [hunk 1]
```diff
@@ -635,13 +635,70 @@ test "lessThan" {
 /// Compares two slices and returns whether they are equal.
 pub fn eql(comptime T: type, a: []const T, b: []const T) bool {
     if (a.len != b.len) return false;
-    if (a.ptr == b.ptr) return true;
+    if (a.len == 0 or a.ptr == b.ptr) return true;
+
+    // No vectorization in stage2 x86_64 or x86
+    // https://github.com/ziglang/zig/issues/17748
+    if (builtin.zig_backend != .stage2_x86_64 or builtin.zig_backend != .stage2_x86) {
```
Comment (squeek502) [inline lib/std/mem.zig:642]:
```suggestion
    if (builtin.zig_backend != .stage2_x86_64 and builtin.zig_backend != .stage2_x86) {
```

Comment (Rexicon226) [reply lib/std/mem.zig:642]: :grief: thank you kind sir

File: lib/std/mem.zig [hunk 2]
```diff
@@ -635,13 +635,70 @@ test "lessThan" {
 /// Compares two slices and returns whether they are equal.
 pub fn eql(comptime T: type, a: []const T, b: []const T) bool {
     if (a.len != b.len) return false;
-    if (a.ptr == b.ptr) return true;
+    if (a.len == 0 or a.ptr == b.ptr) return true;
+
+    // No vectorization in stage2 x86_64 or x86
+    // https://github.com/ziglang/zig/issues/17748
+    if (builtin.zig_backend != .stage2_x86_64 and builtin.zig_backend != .stage2_x86) {
+        if (@typeInfo(T) == .Int and std.math.isPowerOfTwo(@bitSizeOf(T))) return eqlBytes(std.mem.sliceAsBytes(a), std.mem.sliceAsBytes(b));
```
Comment (Vexu) [inline lib/std/mem.zig:643]: Could this be expanded to use `std.meta.hasUniqueRepresentation` (and `@sizeOf(Elem) > 0`)?

Comment (Rexicon226) [reply lib/std/mem.zig:643]: I did not know that was a function, an interesting idea. It seems to do what we want. Do I understand correctly that it ensures `sliceAsBytes` will be a "perfect" slice? If so, then I agree this is a much better suited function for this purpose. I was exploring making things like packed structs work, but got a bit side-tracked.

Comment (Rexicon226) [reply lib/std/mem.zig:643]: To add to that, `@sizeOf(Elem)` being 0 should actually be made a immediate `return true`, as there being no data means there is nothing to compare.

Comment (Rexicon226) [reply lib/std/mem.zig:643]: So, I thought about it some more, and `hasUniqueRepresentation` is the function we want to use. Will finish testing a push shortly. Also to my comment about `u0` not being uniquely represented, I take that back. It doesn't have any padding nor unused bits so it technically fits the description.

File: lib/std/mem.zig [hunk 3]
```diff
@@ -635,13 +635,79 @@ test "lessThan" {
...
+            var x = @as(V, a[0 .. n / 2].*) ^ @as(V, b[0 .. n / 2].*);
+            x |= @as(V, a[a.len - n / 2 ..][0 .. n / 2].*) ^ @as(V, b[a.len - n / 2 ..][0 .. n / 2].*);
+            return @reduce(.Or, x) == 0;
+        }
+    }
+    // Compare inputs in chunks at a time (excluding the last chunk).
+    for (0..(a.len - 1) / Scan.size) |i| {
+        const a_chunk: Scan.Chunk = @bitCast(a[i * Scan.size ..][0..Scan.size].*);
+        const b_chunk: Scan.Chunk = @bitCast(b[i * Scan.size ..][0..Scan.size].*);
+        if (!Scan.isZero(a_chunk ^ b_chunk)) return false;
```
Comment (IntegratedQuantum) [inline lib/std/mem.zig:702]: Is there a reason why you do
```zig
!(@reduce(.Or, a_chunk ^ b_chunk) == 0)
```
instead of
```zig
@reduce(.Or, a_chunk != b_chunk)
```
here?

Looking at [godbolt](https://godbolt.org/z/857nnfqEM), both versions produce the exact same code in release.
However the second version does seem better at conveying the intent here and it also seems to produce less code in debug mode, since the boolean Or reduction can be done in a single instruction on x86, whereas the bitwise Or reduction does take log(n) operations.

File: lib/std/mem.zig [hunk 4]
```diff
@@ -635,13 +635,79 @@ test "lessThan" {
...
     for (a, b) |a_elem, b_elem| {
         if (a_elem != b_elem) return false;
     }
     return true;
 }
 
+/// std.mem.eql heavily optimized for slices of bytes.
+pub fn eqlBytes(a: []const u8, b: []const u8) bool {
+    if (a.len != b.len) return false;
+    if (a.len == 0 or a.ptr == b.ptr) return true;
```
Comment (karlseguin) [inline lib/std/mem.zig:656]: For my own curiosity, I'm wondering how you judged that this optimization was worth keeping?  On the one hand, it's very cheap. On the other, I want to say it'll rarely be true.

Comment (Rexicon226) [reply lib/std/mem.zig:656]: Which one? You selected 3 different ones.

Comment (karlseguin) [reply lib/std/mem.zig:656]: Both the `a.len == 0` and the `a.ptr == b.ptr`

File: lib/std/mem.zig (hunk 3 above)
Comment (Rexicon226) [reply lib/std/mem.zig:702]: Huh, interesting. There is no particular reason for it other than I didn't think it would affect it, but looking at the godbolt I tend to agree with you. It seems the reduce is emitting much cleaner code because of the better defined intent as you say. I have no problem changing it :)

File: lib/std/mem.zig (hunk 4 above)
Comment (Rexicon226) [reply lib/std/mem.zig:656]: Sorry, I accidently started a review last time, missclicked.

TLDR again:
It costs virtually nothing, a mere 2 cmp instructions.
Prevents large cost from happening when it can be exited within the first couple of cycles.
I would not call `a.len == 0` to be rare, it's definitely something that can happen. `a.ptr == b.ptr` is something that also can happen and is a useful check to make. It could be some sort of overlap happening, who knows what. 

There's just no point *not* adding them. They cost virtually nothing.

File: lib/std/mem.zig (hunk 3 above)
Comment (Rexicon226) [reply lib/std/mem.zig:702]: After a couple more performances tests to compare them, I see absolutely no difference between the two methods. Neither in ReleaseFast nor in Debug. I do not see the point of changing it now to be honest. The CI has run for a while, and it would be unfortunate to re-run it for no actual gain. To be honest, the code seems more ungainly now, then before. `isZero` just makes more sense when reading the flow.g

I will continue testing tomorrow, and if I find something that increases performance, I'll be sure to add it :smile:.

File: lib/std/mem.zig (hunk 4 above)
Comment (mlugg) [reply lib/std/mem.zig:656]: You can't measure the cost of something just by analyzing the number of instructions involved - but even if you could, it's not just the `cmp`, as there's also a conditional branch involved here, which is the real source of potential trouble.

In this case, the optimizer may make the wrong choice for how to construct this branch (does "taking the branch" enter the `if`, or skip it?) and lead to a non-trivial cycle count, particularly if mispredicted.

If Zig had likely/unlikely hints, the cost of this code would indeed be very little, as a branch not taken costs a tiny fraction of a cycle. Currently, we do not have such hints. Personally, I would suggest keeping this check, but putting a comment referencing the accepted (but not yet implemented) proposal for an `@expect` builtin.

Comment (Rexicon226) [reply lib/std/mem.zig:656]: Of course I understand that the pure instructions are not the only factor. I might have no mentioned it in my initial reply, but I meant to say there is no actual performance difference. While I understand what you mean about the optimizer and branching, in the end looking at the raw numbers shows no noticeable performance change.

Even then, I just sort of disagree. Looking at the instructions is one of the only ways to tell if the cost will truly matter. The optimizer has finished working on that specific code. There's no branch prediction or missing happening, or question about it. If I had said I was looking at the Air or llvm IR, then obviously this is in a non-optimized state. But objdumping the binary itself does yield the final instructions, and I believe is an accurate way to quickly measure cost.

Comment (Rexicon226) [reply lib/std/mem.zig:656]: While an expected keyword would be massive for the language, and something I desperately desire it doesn't exist yet. My make do is simply doing raw empirical measurements. Looking at the generated assembly on many different targets, in different modes. 

Ive attempted to mimic the expected builtin by making a separate setCold(true) function for returning, and using that. Doesn't affect code generation, because it's already generating the most optimal code.

File: lib/std/mem.zig (hunk 3 above)
Comment (Rexicon226) [reply lib/std/mem.zig:702]: So, as it turns out, I was given a reason to switch to this bool method 😄 
stage2_x86_64 does not currently support non-bool reduces. Changing it to only use bool reduces keeps the same exact performance, while allowing stage2 to make use of it.

Thank you for the idea!

File: lib/std/mem.zig [hunk 5]
```diff
@@ -635,13 +635,76 @@ test "lessThan" {
...
+    if (std.meta.hasUniqueRepresentation(T)) return eqlBytes(std.mem.sliceAsBytes(a), std.mem.sliceAsBytes(b));
+
     for (a, b) |a_elem, b_elem| {
         if (a_elem != b_elem) return false;
     }
     return true;
 }
 
+/// std.mem.eql heavily optimized for slices of bytes.
+pub fn eqlBytes(a: []const u8, b: []const u8) bool {
```
Comment (sno2) [inline lib/std/mem.zig:650]: > The idea here is to split this api into 2 different functions. The normal eql function, which works on anything, and the much more optimized eqlBytes, which specifically optimizes bytes. eql calls eqlBytes if T is u8.

Is there a reason that this is public?

Comment (Rexicon226) [reply lib/std/mem.zig:650]: It's mainly that I didn't see a reason *not* to make it public :) 

If the user wants to use the eqlBytes function, good on them. It both makes the code easier to read, as the purpose is in the name, and removes an extra parameter that doesn't need to exist. I'm not the typo of person to advocate for squeezing out every character and making the names smaller, but I think it's still easier in the long run.

File: lib/std/mem.zig [hunk 6]
```diff
@@ -635,13 +635,76 @@ test "lessThan" {
 /// Compares two slices and returns whether they are equal.
 pub fn eql(comptime T: type, a: []const T, b: []const T) bool {
     if (a.len != b.len) return false;
-    if (a.ptr == b.ptr) return true;
+    if (a.len == 0 or a.ptr == b.ptr) return true;
+    if (@sizeOf(T) == 0) return true;
+
+    if (std.meta.hasUniqueRepresentation(T)) return eqlBytes(std.mem.sliceAsBytes(a), std.mem.sliceAsBytes(b));
```
Comment (rootbeer) [inline lib/std/mem.zig:641]: Should this test to call over to `eqlBytes()` happen before the len/ptr checks above?  Since those are duplicated in the body of eqlBytes and I think this is a comptime check, vs. those are run-time checks.

File: lib/std/mem.zig (hunk 5 above)
Comment (sno2) [reply lib/std/mem.zig:650]: I'm not sure it is necessary to make it public. Per the Zen:

```
 * Only one obvious way to do things.
```

Making it public means that it is objectively fine to write both `std.mem.eql(u8, a, b)` and `std.mem.eqlBytes(a, b)`. I feel like we're better off just allowing one if they do the exact same thing.

Furthermore, people may decide to manually cast their non-`u8` slices to satisfy `eqlBytes`'s `[]const u8` parameter by manually using `std.mem.toBytes`. Therefore, leaving both options to produce the same behavior could lead to unnecessarily uses of `std.mem.eqlBytes`.

Comment (Rexicon226) [reply lib/std/mem.zig:650]: You bring up a very good point. With this in consideration I agree that it should not be made public, and I'll change that :)

Second point increases my willingness to do it. People might not realize how `sliceAsBytes` behaves with padding, and this'll cause issues I imagine.

Thanks!

File: lib/std/mem.zig (hunk 6 above)
Comment (Rexicon226) [reply lib/std/mem.zig:641]: The duplicate checks are optimized out trivially, however when I do what I said in sno2's review, I don't mind changing the order. There's no reason to, but 🤷 why not.

Comment (Rexicon226) [reply lib/std/mem.zig]: Huh, that caused some unexpected results to be honest 😄

Comment (erikarvstedt) [reply lib/std/mem.zig]: In your last commit, you again moved the comptime checks below the len/ptr checks. What's the reason for this?

Comment (Rexicon226) [reply lib/std/mem.zig]: Whoops, thank you for noticing. I guess it was a conflict between two commits I squashed. Will fix now :heart:

**Review by Vexu — APPROVED** [review]

**Comment by andrewrk** [comment]
Great work! I love to see this kind of collaboration. 

[your next mission...](https://github.com/ziglang/zig/blob/2f8e4347b1f9c2bf093ac48a636576c22359c1ff/lib/compiler_rt/memcpy.zig#L11-L28)
